### DISCUSSION TITLE
Support multiple Directory/Mailbox for custom domain

### DISCUSSION THREAD

**Discussion by nguyenkims** [simple-login/app#415] [Feature Requests]
User can create a directory for their domain and choose the mailbox that owns this directory. By default, a directory belongs to the default mailbox.

Moved from https://trello.com/c/CSBsfCh9/59-support-multiple-directory-mailbox-for-custom-domain

**Comment by IpsmLorem** [comment]
Yes please :)

**Comment by danimalquackers** [comment]
Are there any updates to this (specifically directories on custom domains)? Super excited for it to arrive.

**Reply by nguyenkims** [reply]
We have actually started working on this ticket in the past but it gets paused as other priorities kicked in ... We'll try to get back on it as soon as possible, preferably this month or next month.

**Comment by TtomtomT** [comment]
What is the current status on this? I'm on the verge of starting to use SimpleLogin and this feature is definitely something I want to use.

**Reply by nguyenkims** [reply]
Hey I hope this one is ready next month (in 1-2 weeks). The support for multiple mailbox on custom domain is actually already available so the new version will be to support directory in custom domain.

**Comment by danimalquackers** [comment]
Are there any more updates on this? I haven't been using SL much because I don't want to have to recreate every alias when this feature comes out, still holding out hope.

**Reply by nguyenkims** [reply]
Hi unfortunately due to some unexpected event, we haven't been able to start working on this feature.

**Reply by wgurney** [reply]
+1 for this feature. I have imported 86 aliases. Much better vs manual. 

But for new aliases this feature for custom domains will help. Thanks! SL is a great solution!

**Comment by fazy** [comment]
This is a really important feature that will allow me to use SimpleLogin for real (I have a test domain and am happy to pay while I'm evaluating as this is such a great service). However there are a couple of points:

**Special characters**

When you implement it, please allow other separators for the directory part besides `#`, `/` and `+`. I had too many issues with `+` addresses in the past ever to go back to there. Some websites actively block them, some accidentally block them but the worst ones **accept them at first, and then break later**. A common example is password resets not working. Too many hours of my life wasted because of that last case.

I know they're valid characters in an email address, but `+` doesn't work in the wild and I would equally worry about `/` and `#`.

The only non-alphanumeric characters I'm comfortable with are `.` (used everywhere), `-` (used by Yahoo! or at least used to be, along with my own custom solution used on 1000+ addresses over more than 10 years), and `_` (just a hunch).

**Use case**

I expect there are several use cases, e.g. routing to a different person based on the directory prefix. Mine is simpler but perhaps not uncommon:

1. I have a domain that gets way too much spam as the domain itself is compromised, e.g. I'll get email to `alice@example.com` and `bob@example.com` even if I never used those email addresses.
2. I use a specific format for legitimate email addresses that I give out. For example `prefix-suffix@example.com` where `prefix` is common and `suffix` depends on which org I give the email to. Or more recently something like `prefix.suffix.12345@example.com` where prefix and suffix are as before and 12345 is a random number.

To make catch-all bearable, I would like to be able to set a rule to only accept email if the envelope address matches one of the patterns e.g. `^[a-zA-Z0-9]+\.[a-zA-Z0-9]+.\d+$` (or some simplification to avoid exposing regex like `{alphanumeric}.{alphanumeric}.{numeric}`).

Actually, even just "string starts with `<prefix>` would suffice; the boundary could be any non-alphanumeric character. There are several ways to do this, just don't make me feel the pain of using unusual (albeit valid) characters because things _will_ break in painful and annoying ways.

I mention it here because I think many people might use directories to solve that use case.

**Comment by bukyo** [comment]
+1

I'm definitely looking forward to using directories on my own domain. I can see other things keep trumping development for this, but do we have a rough ETA please?

**Comment by jimjoh** [comment]
An ETA on this would be great.  This feature is actually why I signed up for SimpleLogin almost a year ago.

**Reply by danimalquackers** [reply]
I signed up for this as well, and it's disappointing to see how many times this has been pushed back, it would be such a useful feature.

**Reply by nguyenkims** [reply]
Sorry for the delay, we haven't been able to work on this feature due to unexpected events (bugs, holidays, etc). This feature, once implemented will actually be more powerful than the current "directory". The idea is you can define *auto-create rules* to create aliases on the fly. A *rule* can be a prefix (so in this case equivalent to directory), a suffix, or anything else. Rule can be seen as an extension of the "catch-all" feature.

**Reply by bukyo** [reply]
> Sorry for the delay, we haven't been able to work on this feature due to unexpected events (bugs, holidays, etc). This feature, once implemented will actually be more powerful than the current "directory". The idea is you can define _auto-create rules_ to create aliases on the fly. A _rule_ can be a prefix (so in this case equivalent to directory), a suffix, or anything else. Rule can be seen as an extension of the "catch-all" feature.

That's great to hear that it's still coming. How high on the priority list is this feature? As it has been bumped numerous times already. I'm just wondering if it's almost done and likely to be available by the end of the year? Or if it's a lower priority to you and "you'll get to it when you get to it"?

**Reply by nguyenkims** [reply]
This feature is actually the highest in our list of new features for the email part in SimpleLogin (we are working on the phone number part in parallel) but below other urgencies like bug fixes or maintenance. I hope that we'll get to it at the end of this month.

**Reply by bukyo** [reply]
Thank you. It's good to know you have it at the same priority as your users :)

Really appreciate the update from you, and look forward to using the feature once implemented.

**Comment by nguyenkims** [comment]
This is now possible as we can define a regular expression for auto-create alias on a custom domain. This is actually more flexible than directory as we aren't limited to dot, slash, dash as separators. Using regex is never easy though so this feature is considered beta at the moment.

![2021-09-18 at 19 12](https://user-images.githubusercontent.com/3167780/133897151-a91f4d7f-fa7f-4fc9-a8bb-3b682c450404.png)

**Reply by jimjoh** [reply]
Looks like a great feature.  I don't see "Catch All" in my settings yet, will this be available to test soon?

**Reply by bukyo** [reply]
Thank you for putting this in.

I've found the option under Domains > Details... and I've configured it to work with a simple prefix as per the instructions on the page. It appears to work for only the mail that matches the prefix I set, so thank you very much nguyenkims for getting this in .It helps in the short term for generating simple addresses on the fly, without having to whip out the phone and save an alias before sharing your email address (alias) with someone - very handy for trade fairs and the like.

It appears the regex will not read the or ( | ) character though. Are there plans to incorporate this please?

I have one further request for development:

The advantage of the original Directories is that we could set up multiple directories, each going to a mailbox of our choice. i.e. Handy for all the team as I could set "user1+*" to user1's email, "user2+*" to user2's email, etc. I hope the custom domain auto-alias will soon be able to support multiple mailboxes too. I appreciate in the short term we can just go in to aliases and repoint the alias to forward mail to the relevant email address once the first email is received, but this requires yet another touchpoint for the administrator/user. Are there further plans to expand the directories function for custom domains to generate multiple auto-aliases for custom domains, that point to separate mailboxes?

Please don't let my question for further development take away how excited I am to have the initial beta function put in. This is really great so far!

**Reply by jimjoh** [reply]
> The advantage of the original Directories is that we could set up multiple directories, each going to a mailbox of our choice. i.e. Handy for all the team as I could set "user1+_" to user1's email, "user2+_" to user2's email, etc. I hope the custom domain auto-alias will soon be able to support multiple mailboxes too. I appreciate in the short term we can just go in to aliases and repoint the alias to forward mail to the relevant email address once the first email is received, but this requires yet another touchpoint for the administrator/user. Are there further plans to expand the directories function for custom domains to generate multiple auto-aliases for custom domains, that point to separate mailboxes?

I'm impacted by this too.  I need to auto create alias/directories that go to different mailboxes.  Keeping with the current design it'd be nice if i could have multiple regexs, each pointing to a different mailbox.  For example:
```
regex1\..* => maps to mailbox1
regex2\..* => maps to mailbox2
etc.
```
Thanks for implementing this, but  while the catch all with regex is a cool feature, it really doesn't meet my need.  Either allowing directories to work with custom domains or allowing multiple regexs to go to multiple mailboxes would solve my problem.

**Reply by nguyenkims** [reply]
@bukyo the | is now supported :).

In the next iteration, we'll make sure that a custom domain can have several "rule", each rule has a regex, list of mailboxes, and the `order`: the `order` is to know in what order rules should be evaluated if there are several rules and if there are overlappings in the rules.

**Reply by bukyo** [reply]
That's fantastic, I've just tried it and it's working for my initial selections. Thank you so much @nguyenkims 

I'm looking forward to the next iteration too, and am glad you're looking to add multiple rules, and order.

Thanks again.

**Comment by fazy** [comment]
> This is now possible as we can define a regular expression for auto-create alias on a custom domain. 

This is _awseome_, thanks for implementing this so quickly! I plan to test this out when I get a moment.

For now, one question - does the regex matching support and/or imply the beginning/end anchors? (`^` and `$`)? In other words, in the example for matching `prefix\..*`, will `fooprefix.bar` match? If so, it might be worth updating the example to `^prefix\..*` (or `^prefix\..+` as you probably won't want to match just `prefix.` by itself). I appreciate it's just an example though. :)

**Reply by nguyenkims** [reply]
The regex is fully matched so `fooprefix.bar` won't match `prefix\..*` :)

**Comment by nguyenkims** [comment]
The next version is now ready under the new "Auto Create" page. Don't hesitate to give it a try and let me know if you see any bug/possible improvements.

![2021-09-20 at 18 42](https://user-images.githubusercontent.com/3167780/134040881-45510f4a-bb5a-4065-805b-34c8e0639715.png)

**Reply by danimalquackers** [reply]
Just tested this and absolutely love it, much better than just having plain directories. Is there a reason why the supported character set is so small? For example, I'd like to use `prefix\/.*` in a rule, but it doesn't allow for forward-slashes. I also noticed you can't use character classes, which would make it simpler to define rules, like sending anything matching `prefix[+\-\\\/].*` to a single mailbox. I'm imagining this functionality will continue to improve, but those are a couple use cases it doesn't seem to work with yet.

**Reply by nguyenkims** [reply]
Now all characters are allowed :).

I also added a "debug zone" that allows us to test if a regex works as expected.

**Reply by wgurney** [reply]
This entire solution is now fantastic. Thanks so much SimpleLogin Dev team.  I have tried these multi-character regex `.*[.\+\-\\\/]suffix` and `prefix[.\+\-\\\/].*` using multiple rules.
They work great and love the Debug Zone too!

**Reply by danimalquackers** [reply]
Thank you so much for this. Just set up a multi-character rule, super easy and flexible, just great. The only thing I could imagine making it better would be support for editing the rules and being able to drag to change the order, but the feature itself is absolutely amazing and easy already. Side note, I hope that there's something in place to prevent someone from creating rules during a premium trial so they get unlimited aliases after it's expired...

**Comment by jimjoh** [comment]
Love the new feature.  One more bonus idea.. it'd be nice if the alias description for auto created aliases would say something like:
`auto created by regex rule 1 ".*\.mySuffix"`
so  you could easily see why the alias was created.

**Reply by nguyenkims** [reply]
This is a good idea, added :)

**Reply by jimjoh** [reply]
This is awesome, thanks so much.  Love this feature.  Amazing turnaround time on updates!

**Reply by danimalquackers** [reply]
Agreed, while there was a while where I wasn't sure what was going on, this turnaround time with feedback has been refreshing.

**Reply by bukyo** [reply]
> Love the new feature. One more bonus idea.. it'd be nice if the alias description for auto created aliases would say something like: `auto created by regex rule 1 ".*\.mySuffix"` so you could easily see why the alias was created.

In a future update, could you also add the email address of the sender that triggered the generation of the alias, to the alias description please?

I tend to keep a list of the companies I issued the alias address to, in the description so it would help with my maintenance of this.

**Comment by bukyo** [comment]
This is great @nguyenkims it works beautifully when setting up the new rules and the debug zone is a great idea to help us when writing the rules. 

Are we able to edit the rule once saved? There doesn't seem to be an option, and it looks like we have to copy the regex, delete the rule and set up a new one with amended regex. I appreciate this is more of a nice to have, and probably won't be used all the time, but it would make it easier to adjust more complex rules/orders.

Really appreciate you putting this together as it great to be able to finally set up the auto-create directories for our own domain.

**Reply by nguyenkims** [reply]
> Are we able to edit the rule once saved? 

This feature is indeed not included in the v1 to simplify the implementation. In the next iteration, we'll make sure to add it :).

**Reply by bukyo** [reply]
Fantastic. Thank you.

**Comment by argarner** [comment]
This is an outstanding feature (especially the playground) but, after some extensive testing today, v1's implementation does not match the regex's results in regex101.com. Please see https://github.com/simple-login/app/issues/715 to maybe address in v2 @nguyenkims

**Comment by goran69z** [comment]
This might be a good theme for a new YouTube Video where users that are not familiar with GitHub might get acquainted with this feature.

**Comment by fryorcraken** [comment]
I would love to see this feature as it's missing from both SL and anonaddy.

I have a custom domain name to handle aliasing, I would like to be able to use the same custom domain name and same SL account for my family members but have one directory by family member so they can create aliases on the fly.

**Reply by jimjoh** [reply]
> I have a custom domain name to handle aliasing, I would like to be able to use the same custom domain name and same SL account for my family members but have one directory by family member so they can create aliases on the fly.

I do this, but I use "auto create alias rules" (regexes) instead of directories, it works great.
For example with regexes like this:
```
.*\.kid1
.*\.kid2
```
ANYWORD.kid1@my.domain goes to kid1@real.email
ANYWORD.kid2@my.domain goes to kid2@real.email

**Reply by fryorcraken** [reply]
Omg, this option was well hidden in the menus. Thank you so much.
